waku/v2/protocol/waku_store/waku_store_types.nim

```diff
@@ -113,30 +117,272 @@ type
+  
+  var nextItem = w.first
+  
+  ## Fast forward until we reach the startCursor
+  while nextItem.isOk:
+    if nextItem.value.key == startCursor:
+      # Exit ffd loop when we find the start cursor
+      break
+
+    # Not yet at cursor. Continue advancing
+    nextItem = w.next
+
+  return nextItem
+
+proc rwdToCursor(w: SortedSetWalkRef[Index, IndexedWakuMessage],
+                 startCursor: Index):
+                 SortedSetResult[Index, IndexedWakuMessage] =
+  ## Rewind `w` to start cursor
+  ## TODO: can probably improve performance here with a binary/tree search
+  
+  var prevItem = w.last
+
+  ## Rewind until we reach the startCursor
+  
+  while prevItem.isOk:
+    if prevItem.value.key == startCursor:
+      # Exit rwd loop when we find the start cursor
+      break
+
+    # Not yet at cursor. Continue rewinding.
+    prevItem = w.prev
+  
+  return prevItem
+
+proc fwdPage(storeQueue: StoreQueueRef,
+             pred: QueryFilterMatcher,
+             maxPageSize: uint64,
+             startCursor: Option[Index]):
+            (seq[WakuMessage], PagingInfo, HistoryResponseError) =
+  ## Populate a single page in forward direction
+  ## Start at `startCursor`, or first entry if not defined.
+  ## Page size must not exceed `maxPageSize`
+  ## Each entry must match the `pred`
+  
+  var
+    outSeq: seq[WakuMessage]
+    outPagingInfo: PagingInfo
+    outError: HistoryResponseError
+  
+  var
+    w = SortedSetWalkRef[Index,IndexedWakuMessage].init(storeQueue.items)
+    currentEntry: SortedSetResult[Index, IndexedWakuMessage]
+    lastValidCursor: Index
+  
+  # Find first entry
+  if startCursor.isSome():
+    lastValidCursor = startCursor.get()
+
+    let cursorEntry = w.ffdToCursor(startCursor.get())
+    if cursorEntry.isErr:
+      # Quick exit here if start cursor not found
+      outSeq = @[]
+      outPagingInfo = PagingInfo(pageSize: 0, cursor: startCursor.get(), direction: PagingDirection.FORWARD)
+      outError = HistoryResponseError.INVALID_CURSOR
+      return (outSeq, outPagingInfo, outError)
+    
+    # Advance walker once more
+    currentEntry = w.next
+  else:
+    # Start from the beginning of the queue
+    lastValidCursor = Index() # No valid (only empty) last cursor
+    currentEntry = w.first
+
+  ## This loop walks forward over the queue:
+  ## 1. from the given cursor (or first entry, if not provided)
+  ## 2. adds entries matching the predicate function to output page
+  ## 3. until either the end of the queue or maxPageSize is reached
+  var numberOfItems = 0.uint
+  while currentEntry.isOk and numberOfItems < maxPageSize:
+    if pred(currentEntry.value.data):
+      lastValidCursor = currentEntry.value.key
+      outSeq.add(currentEntry.value.data.msg)
+      numberOfItems += 1
+    currentEntry = w.next
+  w.destroy
+
+  outPagingInfo = PagingInfo(pageSize: outSeq.len.uint,
+                             cursor: lastValidCursor,
```

Comment:
Good point! I'd say this may be the expected behaviour, though, and is similar to [what would have been done previously](https://status-im.github.io/nim-style-guide/libraries.results.html?highlight=cstring#results-librariesresults).
In other words, a client may expect to receive an empty response "if there is no more history" in the queried direction, but needs to be able to re-use the cursor received in that same response to continue querying for history (in case more becomes available) or even by querying other synchronised `store` nodes that may hold longer histories.

Reply:
I think returning the same `cursor` in an empty response is fine, it only needs to be explained in the spec as well.